Makefile

```diff
@@ -191,3 +179,23 @@ endef
 define local_image_exists
 $(shell docker images -q $1 2> /dev/null)
 endef
+
+# arguments:
+# $1 - release image repo, e.g. mesosphere/spark
+# $2 - release image tag, e.g spark-2.4.3-hadoop-2.9-k8s
+# $3 - dev image full name, e.g mesosphere/spark-dev:ab36f1f3691a8be2050f3acb559c34e3e8e5d66e
+define tag_and_push_image
+	$(eval RELEASE_IMAGE_FULL_NAME=$(1):$(2))
+	# check, if specified image already exists to prevent overwrites
+	if [[ -z "$(call remote_image_exists,$(1),$(2))" ]]; then
+		docker pull $(3)
+		docker tag $(3) $(RELEASE_IMAGE_FULL_NAME)
+		echo "Pushing image \"$(RELEASE_IMAGE_FULL_NAME)\""
+		docker push $(RELEASE_IMAGE_FULL_NAME)
+	else
+		echo "Error: image \"$(RELEASE_IMAGE_FULL_NAME)\" already exists, will not proceed with overwrite."; false
```

Comment:
can we use $(error <message>) macros instead?

Reply:
In this case, replacing `echo` with `$(error)` function will not work: make will fail with the error even before the recipe is executed (as explained [here](https://stackoverflow.com/questions/11090023/how-do-i-get-error-to-work-conditionally-in-gnu-make)). It would possibly work if we replace shell `if-else` with `$(if)` function but in my opinion, it will affect readability. Let's discuss it offline.

Reply:
As discussed offline, switching to macros approach will either affect readability or introduce unnecessary code duplication. The current approach seems to be a reasonable tradeoff and is good to go.